paddle/fluid/framework/ir/fuse_adjacent_nodes_pass.cc

```diff
@@ -0,0 +1,480 @@
+         (IsActivation(inside_op_name) && IsElemwise(outside_op_name));
+}
+
+void FuseAdjacentNodesPass::FuseElemwiseAndActivation(
+    const NodePtr cur_op_node,
+    const std::unordered_set<NodePtr> &tobe_fused_nodes,
+    OpDesc *op_desc) const {
+  auto upstream_op_node = *tobe_fused_nodes.begin();
+
+  auto upstream_op_type = upstream_op_node->Op()->Type();
+  auto cur_op_type = cur_op_node->Op()->Type();
+  auto upstream_op_node_in_args = upstream_op_node->Op()->InputArgumentNames();
+  auto upstream_op_node_out_args =
+      upstream_op_node->Op()->OutputArgumentNames();
+  auto cur_op_node_in_args = cur_op_node->Op()->InputArgumentNames();
+  auto cur_op_node_out_args = cur_op_node->Op()->OutputArgumentNames();
+
+  if (IsBackward(cur_op_node, tobe_fused_nodes)) {
+    const std::string op_type = "fused_elemwise_activation_grad";
+    op_desc->SetType(op_type);
+
+    // Set attrs
+    op_desc->SetAttr("functor_list",
+                     std::vector<std::string>({upstream_op_type, cur_op_type}));
+    op_desc->SetAttr("recomputation", false);
+
+    auto out_grad = ::paddle::framework::GradVarName("Out");
+    auto x_grad = ::paddle::framework::GradVarName("X");
+    auto y_grad = ::paddle::framework::GradVarName("Y");
+    auto inter_grad = ::paddle::framework::GradVarName("IntermediateOut");
+
+    if (IsElemwise(cur_op_type)) {
+      // the backward of  Unary(Binary(X, Y))
+      PADDLE_ENFORCE(upstream_op_node_out_args.size() == 1,
+                     "The number of output of %s should be 1.",
+                     upstream_op_type);
+      PADDLE_ENFORCE(cur_op_node_out_args.size() == 2,
+                     "The number of output of %s should be 2.", cur_op_type);
+      PADDLE_ENFORCE(
+          upstream_op_node_in_args.size() == 2 ||
+              upstream_op_node_in_args.size() == 3,
+          "The number of inputs of %s should be 2 or 3, "
+          "if the number is 2, the input is 'Out', 'Out@Grad', "
+          "if the number is 3, the input is 'X', 'Out' and 'Out@Grad'.",
+          upstream_op_node->Op()->Type());
+      PADDLE_ENFORCE(
+          cur_op_node_in_args.size() == 2 || cur_op_node_in_args.size() == 4,
+          "The number of inputs of %s should be 2 or 4, if the number is 2, "
+          "the input variable is `Y`, and `Out@Grad`, if the number is "
+          "4, the input variable is `X`, `Y`, `Out`, `Out@Grad`",
+          cur_op_type);
+
+      op_desc->SetInput("X", {});
+      // bool keep_intermediate = true;
+      if (cur_op_node_in_args.size() == 4) {
+        op_desc->SetInput("X", cur_op_node->Op()->Input("X"));
+      }
+
+      if (upstream_op_node_in_args.size() == 3) {
+        op_desc->SetInput("IntermediateOut",
+                          upstream_op_node->Op()->Input("X"));
+      } else {
+        // If the mem_opt is opened and the number of upstream_op_node_in_args
+        // is 2, the Unary is inplace, so the IntermediateOut has been rewrote
```

Comment:
If there is an edge between intermediate-out and backward op, should mem_opt decide not to inplace?

Reply:
I think we should add a pass to determine whether to keep the intermediate-out, it needs a global consideration. 
fuse_adjacent_nodes_pass only takes care of fusing the adjacent nodes in a local space and makes sure the result is right.